llvm/lib/SYCLLowerIR/GlobalOffset.cpp

```diff
@@ -278,6 +300,40 @@ class GlobalOffset : public ModulePass {
       SmallVector<ReturnInst *, 8> Returns;
       CloneFunctionInto(NewFunc, Func, VMap,
                         CloneFunctionChangeType::GlobalChanges, Returns);
+      // In order to keep the signatures of functions called by the kernel
+      // unified, the pass has to copy global offset to an array allocated in
+      // addrspace(3). This is done as kernels can't allocate and fill the
+      // array in constant address space, which would be required for the case
+      // with no global offset.
+      if (AT == ArchType::AMDHSA) {
+        BasicBlock *EntryBlock = &NewFunc->getEntryBlock();
+        IRBuilder<> Builder(EntryBlock, EntryBlock->getFirstInsertionPt());
+        Type *ImplicitOffsetType =
+            ArrayType::get(Type::getInt32Ty(M.getContext()), 3);
+        Value *OrigImplicitOffset =
+            NewFunc->arg_begin() + (NewFunc->arg_size() - 1);
```

Comment:
It can never happen.
This code works on functions collected by `populateKernels` - which looks in the metadata to find all the kernels that use global offset (using `amdgcn|nvvm.annotations` node). As the offset  is passed through a kernel argument (implicit) each of those kernels is guaranteed to have at least 1 arg.